src/mcl_3dl.cpp

```diff
@@ -830,6 +830,98 @@ class MCL3dlNode
 
         markers.markers.push_back(marker);
       }
+      for (auto &p : pc_particle_beam->points)
+      {
+        int beam_header_id = lroundf(p.intensity);
+        Vec3 pos = e.pos + e.rot * origins[beam_header_id];
+        Vec3 end(p.x, p.y, p.z);
+        int num = (end - pos).norm() / params_.map_grid_min;
+        Vec3 inc = (end - pos) / num;
+        end -= inc;
+        for (int i = 0; i < num - 1; i++)
```

Comment:
Separate raycasting part as a function and reuse it from measurement and visualization.